File: Python/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def __init__(self):
        self.ans = 0

    def maxOne(self, root):
        if root == None:
            return 0
        left = max(self.maxOne(root.left), 0)
        right = max(self.maxOne(root.right), 0)
        self.ans = max(self.ans, left + right + root.val)
        return root.val + max(left, right)

    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0
        self.ans = root.val
        self.maxOne(root)
        return self.ans

File: 2021/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {        
        int maxSum = root->val;                
        down(root, maxSum);
        return maxSum;
    }
    
    int down(TreeNode* root, int& maxSum) {
        if (root == NULL) return 0;
        int left = max(down(root->left, maxSum), 0);
        int right = max(down(root->right, maxSum), 0); 
        
        maxSum = max(left + right + root->val, maxSum);
        return max(left, right) + root->val;
    }
    
};